src/System.Device.Gpio/System/Device/Gpio/Drivers/LibgpiodCapiDriver.cs

```diff
@@ -0,0 +1,275 @@
+                {
+                    throw new GpioChipException($"Error while reading value from pin number: {pinNumber}, error: {Marshal.GetLastWin32Error()}");
+                }
+                return (value == 1) ? PinValue.High : PinValue.Low;
+            }
+            return PinValue.Low;
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            if (_pinNumberToEventHandler.TryGetValue(pinNumber, out LibgpiodDriverEventHandler eventHandler))
+            {
+                eventHandler.ValueFalling -= callback;
+                eventHandler.ValueRising -= callback;
+                if (eventHandler.IsCallbackListEmpty())
+                {
+                    eventHandler.CancellationTokenSource.Cancel();
+                    Interop.gpiod_line_release(eventHandler.PinHandle);
+                    eventHandler.Dispose();
+                    _pinNumberToEventHandler.Remove(pinNumber);
+                }
+            }
+            else
+                throw new InvalidOperationException("Attempted to remove a callback for a pin that is not listening for events.");
+        }
+
+        protected internal override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            int success = -1;
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                if (mode == PinMode.Input)
+                {
+                    success = Interop.gpiod_line_request_input(pin, pinNumber.ToString());
+                }
+                else
+                {
+                    success = Interop.gpiod_line_request_output(pin, pinNumber.ToString(), 0);
+                }
+            }
+            if (success == -1) {
+                throw new GpioChipException($"Error setting pin mode, pin:{pinNumber}, error: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        protected internal override WaitForEventResult WaitForEvent(int pinNumber, PinEventTypes eventType, CancellationToken cancellationToken)
+        {
+            bool timedOut = false;
+            int eventRegistered = Interop.gpiod_ctxless_event_loop($"{BasePath}{GpioChip}", (uint)pinNumber, false, $"pin{pinNumber} listener", ref _defaultTimeOut, null,
+                (int e, uint p, ref timespec t, IntPtr d) =>
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        timedOut = true;
+                        return 1;
+                    }
+                    if (e == (int)eventType + 1)
+                    {
+                        return 1;
+                    }
+                    return 0;
+                }, IntPtr.Zero);
+            if (eventRegistered == 0)
+            {
+                return new WaitForEventResult
+                {
+                    TimedOut = timedOut,
```

Comment:
looks like you can removed timedOut as this will always be false.

Reply:
i thought we should set it true when user defined cancellationToken is expired, what you think?